### PR TITLE
Add Agent view page

### LINKED ISSUE #35: Agent view
Going still further down to the ground level we have the agents themselves. These are representations of the population of our fictional town. Some of the population have EVs which are represented by another agent. They have their own activities which they are doing at any given time (shown in a plot) and participate in the DSR system above where possible. We would like to show where the agents are physically in the town and what they are doing at any given time.
- Map showing agent locations – Essentially an image with dots overlaid at certain locations. Dots resize to represent number of agents
- Agent location on SLD (single line diagram schematic) – As above but a different image
- EV Location on Map
- EV Location on SLD
- Agent activity breakdown – A grid of icons representing agents. Coloured by activity
- EV charging breakdown – As above but EVs
- DSR commands to agents – Line graph?

**Proposed Solution:**
See description in #29 

Aha! Link: https://imperial-college-ict.aha.io/requirements/RSE-140-9

### PR DESCRIPTION
**Pull request by tsmbland**
# Description

Agent view page

- SLD images are done and working with random test data. Using the method from Phil (https://github.com/ImperialCollegeLondon/gridlington-vis/issues/35#issuecomment-1798365576). I'm creating a second SVG and overlaying it on top of the SLD image. Saves the SLD image from being redrawn every update. Just need to add a function to map input data to coordinates - I will set up a new issue
- Map images are done using a similar approach to above. Not sure how they want the dots to be formatted but can easily be changed
- Agent activity and EV charging breakdown figures are done using waffle plots. May need to change the layout a bit, and maybe think about pooling agents instead of a single square for every agent

Close #35 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking, back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Key checklist

- [x] All tests pass (`python -m pytest`)
- [x] Pre-commit hooks run successfully (`pre-commit run --all-files`)

### CODE REVIEW

**Comment by AdrianDAlessandro** [comment]
If we remove the SVGs we should remember to squash and merge when we merge this

File: app/svg.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        df (pd.DataFrame): TODO: Opal or DSR?
+
+    Returns:
+        str: SVG url for direct use by html.Img
+    """
+    # Make up data TODO: replace with function that processes real data
+    nodes = svg_sld.raw.split("<circle")[1:]
+    x_coordinates = [float(c.split('cx="')[1].split('"')[0]) for c in nodes]
+    y_coordinates = [float(c.split('cy="')[1].split('"')[0]) for c in nodes]
+    counts = np.random.randint(0, 100, len(x_coordinates))
```
Comment (AdrianDAlessandro) [inline app/svg.py:141]: Can you abstract this into this function you mention but put in the docstring that it needs to be changed and open an issue for that? Maybe even link that issue in the docstring.

Same for the other function(s) like this.

**Review by AdrianDAlessandro — CHANGES_REQUESTED** [review]
This looks really good. A few changes suggested.

File: app/figures.py [hunk 2]
```diff
@@ -356,3 +356,53 @@ def generate_dsr_commands_fig(df: pd.DataFrame) -> px.line:
...
+    TODO: Possibly switch to a waffle plot if desired?
+
+    Args:
+        df: Opal dataframe
+
+    Returns:
+        Plotly express pie chart
+    """
+    if len(df.columns) == 1:
+        agent_activity_breakdown_fig = px.pie()
```
[the remark this answers is not in the document]
Comment (tsmbland) [reply app/figures.py:381]: I can see there's a separate issue for this so I'll leave this for now, but happy to take on that issue

**Review by AdrianDAlessandro — APPROVED** [review]
This is really good, I'm happy to merge it now

**Pull request merged**